design/protocol-changes/object-diffusion.tex

```diff
@@ -261,49 +262,115 @@ \subsection{CDDL encoding specification}
+server to acquire more object ids; in that case it should use a blocking
+request, in order to avoid polling. The rest of the time, the client might
+prefer non-blocking requests to prevent giving agency to the server for too long
+a time, and being able to download objects in the meantime. The server must
+reply within a small timeout to a non-blocking request, possibly with an empty
+list. A blocking request, on the other hand, waits until at least one object is
+available.
+
+The client can request any batch of objects from the current FIFO in any order.
+The server must respond with all the requested objects.%
+%
+\footnote{This is different from the behaviour of Tx-Submission, that allows the
+server to omit transactions in its response, indicating by doing so that they
+are invalid. This behaviour can be recovered by using transaction {\em options}
+as \argfont{object}s.}
+
+The client must ensure that the total size of the object batch it requests does not
+exceed the protocol's size limits. In order to do so, the $\argfont{objectIds}$ type
+should contain enough information to infer the size of each object, or the protocol
+should exchange objects with fixed size.%
+%
+\footnote{For votes and certificates, we will use the latter. See
+\Cref{sec:certificate-vote-diffusion}.}
+
+\subsubsection{Orchestration}
+
+A downstream node should run an instance of the object diffusion mini-protocol
+for each of its upstream peers, with the client side running on the node, and
+the server side running on its peers.
+
+If the network is well-connected, it means that the node should receive the same object id
+from multiple peers. The choice of which peer(s) from which to actually download the object
+depends on a large variety of parameters, specific to each parameterization of the mini-protocol:
+\begin{itemize}
+  \item the latency/throughput constraints,
+  \item the expected network load,
+  \item the quality of each node-to-node connection.
+\end{itemize}
+
+\subsubsection{Hooks}
+
+As the object diffusion protocol is generic, it should be parameterized with hooks to achieve behaviors specific to each use case.
+
+The server should be parameterized with the following hooks:
+\begin{itemize}
+  \item \argfont{initialise}, consuming \argfont{initialPayload}.
+  \item $\argfont{nextIds}$ to retrieve the ids of objects it can serve, when receiving \MsgRequestObjIdsB{} or \MsgRequestObjIdsNB{}.
+  \item $\argfont{idsToObjects}$ to retrieve objects given their ids, when receiving \MsgRequestObjs{}.
+\end{itemize}
+
+The client should be parameterized with the following hooks:
+\begin{itemize}
+  \item $\argfont{onRecvId}$ to add object ids to the FIFO, and to reward/penalize their peer depending on how much time and/or how many ids they sent (similar to the Limit on Patience in Genesis). \nbacquey{Should we put a link to LoP here, or is it assumed the reader knows what we are talking about?}
```

Comment:
I think LoP-like concerns can be discussed in the implementation-specific sections :+1: (I linked to our Genesis docs in #37)